awx/api/views/__init__.py

```diff
@@ -1418,6 +1418,80 @@ class CredentialCopy(CopyAPIView):
     copy_return_serializer_class = serializers.CredentialSerializer
 
 
+class CredentialExternalTest(SubDetailAPIView):
+    """
+    Test updates to the input values of an external credential before
+    saving them.
+    """
+
+    view_name = _('External Credential Test')
+
+    model = models.Credential
+    serializer_class = serializers.EmptySerializer
+
+    def post(self, request, *args, **kwargs):
+        obj = self.get_object()
+        test_inputs = {}
+        for field_name, value in request.data.get('inputs', {}).items():
+            if value == '$encrypted$':
+                test_inputs[field_name] = obj.get_input(field_name)
+            else:
+                test_inputs[field_name] = value
+        try:
+            obj.credential_type.plugin.backend(None, **test_inputs)
+            return Response({}, status=status.HTTP_202_ACCEPTED)
+        except Exception as exc:
+            return Response({'inputs': str(exc)}, status=status.HTTP_400_BAD_REQUEST)
+
+
+class CredentialInputSourceDetail(RetrieveUpdateDestroyAPIView):
+
+    view_name = _("Credential Input Source Detail")
+
+    model = models.CredentialInputSource
+    serializer_class = serializers.CredentialInputSourceSerializer
+
+
+class CredentialInputSourceList(ListAPIView):
+
+    view_name = _("Credential Input Sources")
+
+    model = models.CredentialInputSource
+    serializer_class = serializers.CredentialInputSourceSerializer
+
+
+class CredentialInputSourceSubList(SubListCreateAttachDetachAPIView):
```

Comment:
It is easier for API clients to interact with global lists like `/api/v2/input_sources/` than with sublists like `/api/v2/credentials/N/input_sources/`. Creating one via the global list only means they have to provide both the source and target credential ids. Using the sub-list means they have to know which one is the "parent" resource, and this isn't always obvious to the person writing integrations, it's a distinction they would prefer to not have to think about. For that reason, I advocate introducing a global list as well. If there's any endpoint we don't need, I think it's the sublist.

Reply:
_Maybe_?  We sort of made the decision (for the sake of simplicity) to pick _only_ one of these, and the associate/disassociate model made sense to us here.  I also don't love adding _another_ top level resource here, especially considering this one is *so* specific to the notion of "get Credential `A`'s field value `Y` from Credential `B`".

Reply:
> I also don't love adding _another_ top level resource at `/api/v2/`, especially considering this one is _so_ specific to the notion of "get Credential `A`'s field value `Y` from Credential `B`".

Whenever @matburt talked about this, the principle was always that we make our API "as flat as possible". That's not an actual quote, but I hope I capture the spirit.

So I would respond saying almost the exact rhetorical opposite - `/api/v2/` is tremendously valuable as a comprehensive list of the resources in the system. If a model is not represented there it creates an unwelcome surprise. When the 1 other model I know of which wasn't listed in the top-level endpoint was written into tower-cli, we did very ugly hacks to work it in.

> the associate/disassociate model made sense to us here because the action you're taking is almost exclusively _linking_ and _unlinking_ two things

I agree completely with you for all of the "through" models. For instance, I do like that the group-host and group-children models are not exposed as top-level items. An even better example would be `RoleAncestorEntry`. We define this model explicitly and it has extra fields, but these are never exposed to API clients.

But this `CredentialInputSource` model has user-editable fields. For that matter, job templates are just expressing a relationship between projects and inventories (with a few extra user-editable fields).

Reply:
Should we support _both_, then?  Because I can see the value in both, honestly.

Reply:
> What if `/api/v2/credential_input_sources/` was just a read-only list?

I think that's the worst option of all. I would rather it not be added than to do that. The only thing I was trying to comment on is consistency. For the user, this means consistency of all resource types across the entire app. If this is merged with me having the minority opinion on this, then that's still fine.

Let me return to some other things here. The view class you're using is `SubListCreateAttachDetachAPIView`. I don't think that's entirely right.

How would you attach or detach here? Does that make any sense? When I read your RBAC logic, it implied that `target_credential` could never be null. If it can't be null, then how can you detach? Perhaps you want a sublist-create-API-view? There are 2 categories of sublists - many-to-many relationships and reverse foreignkeys, and they should generally inherit from different base classes. You can still do the creation for many-to-many views, but attach-detach doesn't make sense with reverse foreignkeys. If you were to detach or replace (replace, since this is non-nullable) a foreignkey relationship, that's a PATCH, not a POST {"id": 42} type pattern.

Reply:
Anyway, it looks like the global list is implemented as read-only with `CredentialInputSourceList`

Reply:
@AlanCoding 

The disassociate in this test seems to work fine for me:

https://github.com/ansible/awx/pull/3098/files#diff-731de063081413220909cdc9f7bc3447R38

Am I missing something?

Reply:
I think that test articulates exactly why you shouldn't be able to detach here.

The "disassociate" request is made.

```python
response = post(sublist_url, params, admin)
```

After that request completes, no credential input sources exist

```python
assert CredentialInputSource.objects.count() == 0
```

The act of disassociating the credential input source deleted the credential input source.

I assume this is due to the cascade delete behavior, which is what I would have expected. However, no sane user would ever expect this. If you want to delete it, you should use a DELETE request.

Reply:
Okay, I've added a commit which should address this feedback.